pkg/catalog/retry.go

```diff
@@ -0,0 +1,74 @@
+package catalog
+
+import (
+	"github.com/openservicemesh/osm/pkg/apis/policy/v1alpha1"
+	"github.com/openservicemesh/osm/pkg/identity"
+	"github.com/openservicemesh/osm/pkg/policy"
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+// GetRetryPolicy returns the RetryPolicySpec for the given downstream identity and upstream service
+// If a source and destination has more than one retry policy that can be applied, the more
+// specific retry policy will be used. Types of retry policy:
+// - specific retry policy that specifies the source and destination
+// - wildcard destination service retry policy
+// - wildcard destination service and namespace retry policy
+func (mc *MeshCatalog) GetRetryPolicy(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService) *v1alpha1.RetryPolicySpec {
+	if !mc.configurator.GetFeatureFlags().EnableRetryPolicy {
+		log.Trace().Msgf("Retry policy flag not enabled")
+		return nil
+	}
+	src := downstreamIdentity.ToK8sServiceAccount()
+
+	// List the retry policies for the source
+	retryPolicies := mc.policyController.ListRetryPolicies(src)
+	if retryPolicies == nil {
+		log.Trace().Msgf("Did not find retry policy for downstream service %s", src)
+		return nil
+	}
+	var retry *v1alpha1.RetryPolicySpec
+	prioritizeRetry := false
+	for _, retryCRD := range retryPolicies {
+	forLoop:
+		for _, dest := range retryCRD.Spec.Destinations {
+			if dest.Kind == policy.RetrySourceKindSvc {
+				destMeshSvc := service.MeshService{Name: dest.Name, Namespace: dest.Namespace}
+				switch {
+				case checkSameSvc(upstreamSvc, destMeshSvc):
+					// Will return retry policy that applies to the specific upstream service
+					return &retryCRD.Spec.RetryPolicy
+
+				case dest.Name == identity.WildcardServiceIdentity.String() && upstreamSvc.Namespace == dest.Namespace:
+					// Will use retry policy that applies to all upstream services in destination namespace unless a more specific retry policy is found
+					retry = &retryCRD.Spec.RetryPolicy
+					prioritizeRetry = true
+					break forLoop
+
+				case !prioritizeRetry && dest.Name == identity.WildcardServiceIdentity.String() && dest.Namespace == identity.WildcardServiceIdentity.String():
+					// Will use retry policy that applies to all upstream services for source unless a more specific retry policy is found
+					retry = &retryCRD.Spec.RetryPolicy
+					break forLoop
+				}
+			} else {
+				log.Error().Msgf("Retry policy destinations must be a service: %s is a %s", dest, dest.Kind)
+			}
+		}
+	}
+
+	if !isRetryEmptyOrNil(retry) {
+		return retry
+	}
+	log.Trace().Msgf("Could not find retry policy for source %s and destination %s", src, upstreamSvc)
+	return nil
```

Comment:
Could you simply return the retry object and let the caller decide how it wants to handle nil values? I would suggest setting a nil value for Retry in the route objects, so regardless of whether a non-nil or nil value is returned here ideally should not make a difference.